{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RA single cell Project: preprocessing and analysis\n",
    "\n",
    "<br>\n",
    "\n",
    "**Description**:This Jupyternotebook summarise the single cell pre-processing, the batch correction (Harmonypy) and leiden clustering, the cell subset annotations for the PBMC and each subset\\\n",
    "**Author**: Brenda Miao - Marie Binvignat\\\n",
    "**Version**: 1.0  \n",
    "**Last updated**: 02/01/23\n",
    "\n",
    "<br>\n",
    "\n",
    "**<font size=\"2\">Single Cell Transcriptomics Analysis of Blood Immune Cells Identifies Gene Signatures and Cell Subsets Associated with Disease Activity in a Diverse Population of Rheumatoid Arthritis Patients<font>**\n",
    "<br>  \n",
    "<font size=\"2\"> Binvignat M*, Miao B*, Wibrand C*,Yang M, Rychkov D, Flynn E, Nititham J, Khan U, Tamaki W, Carvidi A, Krueger M, Niemi E, Sun Y, Fragiadakis G, Klatzmann D, Sellam J, Mariotti-Ferrandiz E, Gross A, Ye J, Butte AJ, Criswell LA, Nakamura M, Sirota M\n",
    " <font><br>\n",
    "<font size=\"1\"> **equal contribution* <font> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load functions\n",
    "\n",
    "Loading python pacakges, preprocessing, helping afunctions used for the analysis\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## start here\n",
    "# get functions\n",
    "from functions.preprocessing import *\n",
    "from functions.DEGAnalysis import *\n",
    "\n",
    "#sns.set_style(\"ticks\")\n",
    "#sns.set_context(\"talk\")\n",
    "\n",
    "sns.set(style=\"ticks\")\n",
    "sns.set_context(\"notebook\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reading in the data\n",
    "\n",
    "The dataset consisted of 10X single cell RNA sequencing data from PBMC samples taken from 18 early rheumatoid arthritis patients and 18 controls matched by age, sex and ethnicity. Sequencing was performed in 3 batches, with RA and control samples split evenly within batch and lanes. Sample deconvolution and doublet identification was performed using demuxlet and seurat prior to import into scanpy using count, index/column name, and metadata files.\n",
    "\n",
    "\n",
    "\n",
    "#### Reference for DAS 28 CRP\n",
    "\n",
    "https://www.rheumatology.org/Portals/0/Files/2019-Update-ACR-Recommended-Rheumatoid-Arthritis-Disease-Activity-Measures.pdf\n",
    "\n",
    "**2019 update of the American College of Rheumatology recommended rheumatoid arthritis \n",
    "disease activity measures**.\\\n",
    "England BR, Tiong BK, Bergman MJ, Curtis JR, Kazi S, Mikuls TR, O'Dell JR, Ranganath VK, Limanni A, Suter LG, Michaud K.  \\\n",
    "*Arthritis care & research. 2019 Dec;71(12):1540-55.*\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### params\n",
    "adataX = './dataInput/seurat_demux.mtx.gz'\n",
    "genes = './dataInput/seurat_genes.csv'\n",
    "barcodes = './dataInput/seurat_barcodes.csv'\n",
    "metadata = './dataInput/seurat_demux_meta.csv'\n",
    "\n",
    "outputPath = './dataOutput/adata/seurat_demux.h5ad'\n",
    "\n",
    "### converting from seurat object to h5ad object (takes a while so only run once)\n",
    "adata = loadFromSeurat(adataX, genes, barcodes, metadata)\n",
    "\n",
    "adata.write(outputPath)\n",
    "\n",
    "\"\"\"### Generating metadata file\n",
    "\n",
    "# read in separate files\n",
    "outputPath = './dataInput/patientData/RA_metadata_merged.csv'\n",
    "metadata_RA = pd.read_csv(\"./dataInput/patientData/RA_patient_DAS_MaryV2.csv\")\n",
    "metadata_demographics = pd.read_csv(\"./dataInput/patientData/RA_patient_demographics_MaryV2.csv\")\n",
    "metadata_mapping = pd.read_csv(\"./dataInput/patientData/RA_singlecell_mapping.csv\")\n",
    "\n",
    "# preprocess\n",
    "metadata_demographics = metadata_demographics.dropna(how = 'all')\n",
    "del metadata_demographics[\"dx\"]\n",
    "del metadata_demographics[\"race\"]\n",
    "\n",
    "# merge and save\n",
    "meta_test = metadata_mapping.merge(metadata_RA,left_on = \"ptnum\", right_on = \"sample_id\", how = \"left\")\n",
    "meta_df = meta_test.merge(metadata_demographics,right_on = \"subject_id\", left_on = \"subject_id\", how = \"left\")\n",
    "\n",
    "# drop duplicates\n",
    "#meta_df = meta_df[~meta_df[\"sample\"].duplicated()]\n",
    "meta_df.to_csv(outputPath)\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocessing\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### params\n",
    "inputPath = './dataOutput/adata/seurat_demux.h5ad'\n",
    "outputPath = './dataOutput/adata/normalized.h5ad'\n",
    "plt.rcParams['figure.figsize'] = 4,4\n",
    "seed = 123\n",
    "np.random.seed(seed)\n",
    "\n",
    "# load integrated data\n",
    "adata = sc.read(inputPath, cache=False)\n",
    "\n",
    "# clean up data and add in metrics for each cohort\n",
    "adata = adata[adata.obs[\"demux.doublet.call\"]==\"SNG\",:]\n",
    "adata = cleanRAData(adata, plot = True)\n",
    "\n",
    "# Removing AS patients RA02 with too few cells (68 only) and matching control\n",
    "adata = adata[adata.obs[\"sampleType\"] != \"AS\"]\n",
    "adata = adata[~adata.obs[\"Sample\"].isin([\"RA02_1_C_F\"])] \n",
    "adata = adata[~adata.obs[\"Sample\"].isin([\"Control02_1_C_F\", \"Control11_1_C_M\", \"Control12_1_C_F\",\"Control13_1_C_M\", \"Control14_1_C_F\"])]\n",
    "\n",
    "# Remove platelets (> 1 count of platelet/megakaryocyte gene markers (PF4, SDPR, PPBP, GNG11 (not in dataset)))\n",
    "plt_genes = sc.get.obs_df(adata, keys=[\"PF4\", \"GNG11\", \"PPBP\"])\n",
    "adata = adata[adata.obs.index.isin(plt_genes[plt_genes.sum(axis=1) < 1].index)]\n",
    "\n",
    "# add metrics for QC and preprocess according to thresholds\n",
    "adata = plotQCMetrics(adata, groupby=\"Lane\")\n",
    "adata = set_filters(adata, min_cells=3, min_genes=100, max_genes=1000, min_count=0, max_count=2000, \n",
    "               mtpct=20, rbpct=3, remove_rb=False, remove_hb=False, remove_mt=False) \n",
    "\n",
    "adata = adata[adata.obs.sort_values(\"demux_RD_PASS\").index]\n",
    "\n",
    "# Normalize and log-scale. Store raw counts in adata.raw\n",
    "# NOTE: scaling & regress out performed after select HVG in later cell\n",
    "adata.raw = adata\n",
    "sc.pp.normalize_total(adata, target_sum=1e4, key_added=\"norm_library_size\")\n",
    "sc.pp.log1p(adata) \n",
    "\n",
    "# QC plotting\n",
    "sc.pl.highest_expr_genes(adata, n_top=20);\n",
    "\n",
    "## Cell cycle scoring\n",
    "#Remove s and g2m gene not in our dataset ['MLF1IP', 'FAM64A', 'HN1']\n",
    "s_genes_list = [\"MCM5\",\"PCNA\", \"TYMS\",\"FEN1\", \"MCM2\",\"MCM4\",\"RRM1\",\"UNG\",\"GINS2\",\"MCM6\",\"CDCA7\",\"DTL\",\"PRIM1\",\"UHRF1\", \n",
    "                \"HELLS\",\"RFC2\",\"RPA2\",\"NASP\",\"RAD51AP1\",\"GMNN\",\"WDR76\",\"SLBP\",\"CCNE2\",\"UBR7\",\"POLD3\",\"MSH2\",\n",
    "                \"ATAD2\", \"RAD51\",\"RRM2\",\"CDC45\",\"CDC6\",\"EXO1\",\"TIPIN\",\"DSCC1\",  \"BLM\",\"CASP8AP2\",\"USP1\",\"CLSPN\",\"POLA1\",\n",
    "                \"CHAF1B\",\"BRIP1\" ,\"E2F8\"]\n",
    "\n",
    "g2m_genes_list = [\"HMGB2\",\"CDK1\",\"NUSAP1\",\"UBE2C\",\"BIRC5\",\"TPX2\",\"TOP2A\",\"NDC80\",\"CKS2\",\"NUF2\",\"CKS1B\",\"MKI67\",\"TMPO\",\n",
    "             \"CENPF\",\"TACC3\",\"SMC4\",\"CCNB2\",\"CKAP2L\", \"CKAP2\",\"AURKB\",\"BUB1\",\"KIF11\",\"ANP32E\",\"TUBB4B\",\n",
    "             \"GTSE1\",\"KIF20B\",\"HJURP\",\"CDCA3\",\"CDC20\",\"TTK\",\"CDC25C\",\"KIF2C\",\"RANGAP1\",\"NCAPD2\",\n",
    "             \"DLGAP5\",\"CDCA2\",\"CDCA8\",\"ECT2\",\"KIF23\",\"HMMR\",\"AURKA\",\"PSRC1\",\"ANLN\",\"LBR\",\"CKAP5\",\"CENPE\",\"CTCF\",\n",
    "                  \"NEK2\",\"G2E3\",\"GAS2L3\",\"CBX5\",\"CENPA\"]\n",
    "\n",
    "sc.tl.score_genes_cell_cycle(adata = adata, s_genes = s_genes_list,g2m_genes = g2m_genes_list )\n",
    "adata.obs[\"cell_cycle_diff\"] = adata.obs[\"S_score\"] - adata.obs[\"G2M_score\"]\n",
    "\n",
    "### add metadata\n",
    "RA_meta = pd.read_csv(\"./dataInput/patientData/RA_metadata_merged_221118.csv\", index_col=0, encoding=\"ISO-8859-1\")\n",
    "\n",
    "adata.obs = adata.obs[[c for c in adata.obs.columns if c not in RA_meta.columns]]\n",
    "adata.obs = adata.obs.reset_index(drop=False)\n",
    "\n",
    "RA_meta['group']= RA_meta['group'].apply(str)\n",
    "RA_meta[\"sample_group\"] = RA_meta[\"sample\"]+ RA_meta[\"group\"]\n",
    "adata.obs[\"sample_group\"] = adata.obs[\"Sample\"].astype(str) + adata.obs[\"batch\"].astype(str)\n",
    "adata.obs = adata.obs.merge(RA_meta, how=\"left\", left_on=\"sample_group\", right_on=\"sample_group\")\n",
    "adata.obs = adata.obs.set_index(\"barcodes\", drop=True)\n",
    "\n",
    "# formattting to make adata happy\n",
    "adata.obs[\"mtx\"] = adata.obs[\"mtx\"].astype(str)\n",
    "adata.obs[\"tnf\"] = adata.obs[\"tnf\"].astype(str)\n",
    "adata.obs[\"early_ra\"] = [\"Control\" if x==\"Control\"\n",
    "                         else \"True\" if y==True \n",
    "                         else \"False\" for x,y in zip(adata.obs[\"sampleType\"],adata.obs[\"early_ra\"])]\n",
    "adata.obs[\"acr2010\"] = [\"Control\" if x==\"Control\"\n",
    "                         else \"True\" if y==True \n",
    "                         else \"False\" for x,y in zip(adata.obs[\"sampleType\"],adata.obs[\"acr2010\"])]\n",
    "adata.obs[\"acr1987\"] = [\"Control\" if x==\"Control\"\n",
    "                         else \"True\" if y==True \n",
    "                         else \"False\" for x,y in zip(adata.obs[\"sampleType\"],adata.obs[\"acr1987\"])]\n",
    "\n",
    "# label values\n",
    "adata.obs[\"activity_python_crp\"] = [\"control\" if d==\"control\"\n",
    "                                else \"NA\" if np.isnan(r)\n",
    "                                else \"Remission\" if r<2.6\n",
    "                               else \"Low\" if r<3.2\n",
    "                               else \"Moderate\" if r<=5.1\n",
    "                               else \"High\" for d, r in zip(adata.obs[\"dx\"], adata.obs[\"MY_das28crp4\"])]\n",
    "\n",
    "adata.obs[\"activity_python_binary_crp\"] = [\"control\" if d==\"control\"\n",
    "                                       else \"NA\" if \"NA\" in r\n",
    "                                else \"No\" if \"Remission\" in r\n",
    "                               else \"No\" if \"Low\" in r\n",
    "                               else \"Yes\" if \"Moderate\" in r\n",
    "                               else \"Yes\" for d, r in zip(adata.obs[\"dx\"], adata.obs[\"activity_python_crp\"])]\n",
    "\n",
    "# label values\n",
    "adata.obs[\"activity_python_esr\"] = [\"control\" if d==\"control\"\n",
    "                                else \"NA\" if np.isnan(r)\n",
    "                                else \"Remission\" if r<2.6\n",
    "                               else \"Low\" if r<3.2\n",
    "                               else \"Moderate\" if r<=5.1\n",
    "                               else \"High\" for d, r in zip(adata.obs[\"dx\"], adata.obs[\"MY_das28esr4\"])]\n",
    "\n",
    "adata.obs[\"activity_python_binary_esr\"] = [\"control\" if d==\"control\"\n",
    "                                       else \"NA\" if \"NA\" in r\n",
    "                                else \"No\" if \"Remission\" in r\n",
    "                               else \"No\" if \"Low\" in r\n",
    "                               else \"Yes\" if \"Moderate\" in r\n",
    "                               else \"Yes\" for d, r in zip(adata.obs[\"dx\"], adata.obs[\"activity_python_esr\"])]\n",
    "\n",
    "\n",
    "adata.obs[\"smoker_any\"] = [\"control\" if c==\"control\"\n",
    "                           else \"Unknown\" if type(m)!=str\n",
    "                           else \"No\" if \"Never\" in m\n",
    "                           else \"Yes\" for c,m in zip(adata.obs[\"dx\"], adata.obs[\"smoke_status\"]) ]\n",
    "\n",
    "adata.obs[\"age_group\"] = [\"Unknown\" if m==np.nan\n",
    "                          else \"Old\" if m > 45\n",
    "                          else \"Young\" for m in adata.obs[\"age\"]]\n",
    "adata.obs[\"age_group\"].value_counts()\n",
    "\n",
    "# print final metrics\n",
    "print('\\nCELLS x GENES = ' + str(adata.shape) + '\\n')\n",
    "    \n",
    "print('METRICS FOR RAW DATA: \\n')\n",
    "printMetrics(adata, [\"sampleType\", \"sex\", \"Lane\", \"Sample\"])\n",
    "\n",
    "\n",
    "# normalized adata has raw values in .raw and log-norm values in .X\n",
    "adata.write(outputPath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Batch correction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "### params\n",
    "inputPath = './dataOutput/adata/normalized.h5ad'\n",
    "outputPCs = \"./dataOutput/harmony_PCs.csv\"\n",
    "plt.rcParams['figure.figsize'] = 4,4\n",
    "seed = 123\n",
    "np.random.seed(seed)\n",
    "\n",
    "### load integrated data\n",
    "adata = sc.read(inputPath, cache=False)\n",
    "\n",
    "### Subset to highly variable genes, justification for values\n",
    "print(adata.var[\"mean_counts\"].min())\n",
    "print(adata.var[\"mean_counts\"].mean())\n",
    "print(adata.var[\"mean_counts\"].max())\n",
    "\n",
    "# select HVG \n",
    "adata.uns['log1p']['base'] = None\n",
    "sc.pp.highly_variable_genes(adata, flavor='seurat', min_mean=0.01, max_mean=5, min_disp=0.5,batch_key='batch')\n",
    "sc.pl.highly_variable_genes(adata)\n",
    "adata = adata[:, adata.var.highly_variable]\n",
    "print(\"Highly variable genes: %s\"%len(adata.var_names))\n",
    "\n",
    "# regress out unwanted variables\n",
    "sc.pp.regress_out(adata, ['total_counts', \"n_genes_by_counts\", 'pct_counts_mt','pct_counts_ribo', 'cell_cycle_diff'])\n",
    "\n",
    "# scale data\n",
    "sc.pp.scale(adata, zero_center = True, max_value=10) \n",
    "\n",
    "### dimensional reduction and plotting\n",
    "sc.pp.pca(adata, n_comps=50)\n",
    "sc.pl.pca_variance_ratio(adata, log=True, n_pcs = 50)\n",
    "\n",
    "# calculate neighbors and UMAP embeddings\n",
    "sc.pp.neighbors(adata, n_pcs=50, n_neighbors=15)\n",
    "sc.tl.umap(adata, min_dist=0.1, spread=1)\n",
    "\n",
    "# plot UMAP\n",
    "sc.pl.umap(adata, color=['sampleType', 'batch', 'Lane'], wspace=0.35)\n",
    "\n",
    "# run harmonypy (PREVIOUSLY: 4 iterations, before then included \"Lane\")\n",
    "batchCorrect(adata, output_path=outputPCs, max_iter_h=3, vars_use=[\"batch\"]) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clustering and rough and fine subset annotation \n",
    "\n",
    "\n",
    "\n",
    "Samples were clustered in an unsupervised manner using leiden clustering and rough annotations were made using a leiden resolution of 3, each cluster were assigned to CD4 T cells,  CD8 T cells, Monocytes, NK cells, B cells.\n",
    "\n",
    "Marker genes for each cluster were inspected to ensure high fidelity annotations. Platelets, erthryocytes and suspected doublets were removed from further analysis.\n",
    "\n",
    "Doublets and Rbc-Platelets were excluded from the analysis. \n",
    "\n",
    "Subclustering was repeated for each cell type, CD4 T cells, CD8 T cells, B cells, NK cells, and monocytes, with manual annotations based on marker genes and leiden ranked genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cell Dictionnary \n",
    "\n",
    "For each cell subset this dictionnary provide the leiden keys, the marker genes, the fine annotation and the color code "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cells_dict ={\"CD4Tcells\":{\"leiden\": 0.8,\n",
    "                          \"markergenes\": [\"PTPRC\", \"CCR7\", \"IL2RB\", \"SELL\", \"CD27\", \"CD28\",\"IL7R\",\"FOXP3\",\n",
    "                                          \"CTLA4\", \"IL10\", \"LAG3\", \"TGFB1\", \"IL2RB\", \"CD274\",\"CCR7\", \"IL2\", \"FAS\",\n",
    "                                          \"IFNG\", \"IL4\", \"IL5\", \"GATA3\", \"STAT4\", \"TBX21\"],\n",
    "                         \"fine_annot\": {\"0\" :\"CD4 T Naive\", \"1\" :\"CD4 T effector memory\", \"2\" :\"CD4 T central memory\",\n",
    "                                   \"3\" :\"CD4 T central memory\", \"4\" : \"CD4 T central memory\",\"5\" :\"CD4 T Naive\",\n",
    "                                   \"6\" :\"Dead_cells\",\"7\" :\"yd T cells\", \"8\" :\"Doublets\",\"9\" :\"CD4 T central memory\",\n",
    "                                   \"10\" :\"CD4 T IFIT\"},\n",
    "                         \"color_subset\": [ '#ea4e40','#f37761', '#d33028', '#b3302a', '#72322f'],\n",
    "                         \"color_cmap\":\"Reds\"},\n",
    "             \n",
    "             \"CD8Tcells\":{\"leiden\": 0.7,\n",
    "                          \"markergenes\": [\"PTPRC\", \"CCR7\", \"SELL\", \"CD28\", \"IL7R\" , \"PECAM1\", \"TCF7\",\"ITGAL\",\n",
    "                                            \"SPN\", \"CD44\", \"ITGA4\", \"IL2RB\", \"KLRG1\", \"SELL\",\n",
    "                                            \"KLRB1\", \"GZMH\", \"NKG7\", \"FGFBP2\", \"KLRG1\", \"PRF1\", \"FCGR3A\",\"IFNG\", \"GZMB\",\n",
    "                                            \"TIGIT\",  \"TOX\", \"IL23R\", \"RORC\", \"EOMES\", \"TBX21\"],\n",
    "                          \"fine_annot\":{\"0\":\"CD8 TEMRA\",\"1\":\"CD8 T Naive\", \"2\":\"CD8 T early Tem\",\"3\":\"CD8 T Naive\",\"4\":\"Doublets\",\"5\":\"Dead_cells\"},\n",
    "                          \"color_subset\":['#FFD6A5','#d08c60','#99582a'],\n",
    "                          \"color_cmap\":\"Oranges\"},   \n",
    "             \n",
    "             \"Monocytes\":{\"leiden\":0.8,\n",
    "                          \"markergenes\": [\"CD14\",\"IL1B\", \"IFI6\", \"IFITM3\", \"NFKBIA\", \"NLRP3\",\"FCGR3A\",\"CDKN1C\",\n",
    "                                      \"S100A8\", \"C1QA\", \"TNF\", \"IL6\", \"IL1B\", \"ITGAM\"],\n",
    "                          \"fine_annot\":{\"0\" :\"Myeloid DCs\",\"1\" :\"IL1b-Monocytes\",\"2\" :\"Classical Monocytes\",\"3\": \"Classical Monocytes\", \"4\": \"Non-classical Monocytes\",\n",
    "                                  \"5\": \"Platelets_RBCs\",\"6\": \"IFITM3 Monocytes\",\"7\": \"Doublets\",\"8\": \"Doublets\",\"9\": \"Doublets\", \"10\": \"Doublets\", \"11\":\"Myeloid DCs\"},\n",
    "                          \"color_subset\":['#afaed4','#8d89c0','#705eaa', '#7209b7','#3c096c'],\n",
    "                          \"color_cmap\":\"Purples\"},\n",
    "             \n",
    "             \"NKcells\":{\"leiden\":0.5,\n",
    "                        \"markergenes\":[\"NKG7\", \"NCAM1\", \"KLRD1\" , \"NCR1\", \"KLRB1\", \"GNLY\", \"CCR7\", \"NCR1\"],\n",
    "                        \"fine_annot\":{\"0\" :\"NKCD56bright\", \"1\" :\"NKCD56bright\",  \"2\" :\"NKCD56low\", \"3\": \"Doublets\"},\n",
    "                        \"color_subset\":['#8ed08b', '#2c944c'],\n",
    "                        \"color_cmap\":\"Greens\"},\n",
    "             \n",
    "             \"Bcells\":{\"leiden\":0.8,\n",
    "                       \"markergenes\":[\"CD79A\", \"CD19\", \"CD27\", \"CD38\", \"CD7\", \"ITGAM\", \"CD38\", \"MS4A1\", \"ITGA4\", \"IGHD\", \"IGKC\", \"IGHM\"],\n",
    "                       \"fine_annot\": {\"0\" :\"Naive Bcells\", \"1\" :\"Memory Bcells\", \"2\" :\"Naive Bcells\", \"3\": \"Memory Bcells\",\n",
    "                                     \"4\": \"Plasmablasts\",\"5\": \"Plasmablasts\",\"6\": \"Platelets_RBCs\",\"7\": \"Doublets\",\"8\": \"Doublets\"},\n",
    "                       \"color_subset\":['#2a6f97','#4cc9f0','#023e8a','#0267c1',  '#A0C4FF'],\n",
    "                       \"color_cmap\": \"Blues\"}}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Overall Leiden clustering \n",
    "\n",
    "Computing the leiden clustering for the global adata, the leiden resolution used is 3.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "### params\n",
    "pcaPath = \"./dataOutput/harmony_PCs.csv\"\n",
    "inputPath = './dataOutput/adata/normalized.h5ad'\n",
    "outputPath = './dataOutput/adata/clustered.h5ad'\n",
    "plt.rcParams['figure.figsize'] = 4,4\n",
    "leiden_key = \"leiden\"\n",
    "rank_key = \"leiden_markers\"\n",
    "seed = 123\n",
    "np.random.seed(seed)\n",
    "\n",
    "### read input\n",
    "# normalized adata has raw values in .raw and log-norm values in .X\n",
    "adata = sc.read(inputPath, cache=False)\n",
    "adata.uns['log1p']['base'] = None\n",
    "\n",
    "# load corrected PCs\n",
    "PCA_df = pd.read_csv(pcaPath, index_col=0)\n",
    "adata.obsm['X_pca'] = PCA_df.values\n",
    "\n",
    "# calculate neighbors and UMAP embeddings\n",
    "sc.pp.neighbors(adata, n_pcs=50, n_neighbors=15) \n",
    "sc.tl.umap(adata, min_dist=0.1, spread=1)\n",
    "\n",
    "# plot, with normalized data stored in raw values\n",
    "sc.pl.umap(adata, color=['sampleType', 'batch', 'Lane'], wspace=0.35)\n",
    "sc.pl.umap(adata, ncols=4, color=[\"CD2\", \"CD3D\", \"CD8A\",\"CD4\",\n",
    "                         \"CD33\", \"ITGAM\",\"CD14\", \"FCGR3A\",\n",
    "                        \"NKG7\", \"NCAM1\", \"CD79A\",  \"CD19\"], size=3, vmax=5, use_raw=True)\n",
    "\n",
    "### leiden params\n",
    "leiden_key = \"leiden\"\n",
    "rank_key = \"leiden_markers\"\n",
    "groupby = \"leiden_r3.0\"\n",
    "\n",
    "# calculate leiden \n",
    "sc.tl.leiden(adata, resolution=3.0, key_added=leiden_key+\"_r3.0\")\n",
    "\n",
    "# plot leiden & marker genes\n",
    "sc.pl.umap(adata, color=[groupby])\n",
    "\n",
    "### rank genes groups\n",
    "adata.uns['log1p']['base'] = None\n",
    "sc.tl.rank_genes_groups(adata, groupby=groupby, use_raw=True, pts=True, key_added=rank_key, method='wilcoxon')\n",
    "sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False, key=rank_key, use_raw = False, ncols=6)\n",
    "sc.pl.rank_genes_groups_dotplot(adata, n_genes=5, key=rank_key, groupby=groupby, use_raw = False)\n",
    "\n",
    " # histplots\n",
    "plt.rcParams['figure.figsize'] = [40, 5]\n",
    "sc.pl.violin(adata, ['n_genes_by_counts'],jitter=0.1, groupby = groupby, rotation= 90,cut=0) \n",
    "sc.pl.violin(adata, ['pct_counts_mt'],jitter=0.1, groupby = groupby, rotation= 90,cut=0) \n",
    "sc.pl.violin(adata, ['pct_counts_ribo'],jitter=0.1, groupby = groupby, rotation= 90,cut=0) \n",
    "sc.pl.violin(adata, ['pct_counts_hb'],jitter=0.1, groupby = groupby, rotation= 90,cut=0) \n",
    "\n",
    "### save clustered anndat\n",
    "adata.write(outputPath)\n",
    "\n",
    "### sc.pl.umap(adata, color=['sampleType', 'batch', 'Lane'], wspace=0.35)\n",
    "plt.rcParams['figure.figsize'] = 5,5\n",
    "cluster_small_multiples(adata, clust_key=\"leiden_r3.0\", ncols =6)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rough annotation and creation of cell subset adata and leiden clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Compositional analysis on unannotated clusters\n",
    "# params\n",
    "plt.rcParams['figure.figsize'] = [5, 5]\n",
    "inputPath = './dataOutput/adata/clustered.h5ad'\n",
    "outputPath = './dataOutput/adata/roughannot.h5ad'\n",
    "rank_key = \"leiden_markers\"\n",
    "groupby = \"leiden_r3.0\"\n",
    "leiden_key = \"leiden_subset\"\n",
    "seed = 123\n",
    "np.random.seed(seed)\n",
    "\n",
    "# load adata  \n",
    "# normalized adata has raw values in .raw and log-norm values in .X\n",
    "adata = sc.read(inputPath, cache=False)\n",
    "\n",
    "# plotting\n",
    "#sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False, key=rank_key, use_raw = False, ncols=6)\n",
    "#sc.pl.rank_genes_groups_dotplot(adata, n_genes=5, key=rank_key, groupby=groupby, use_raw = False)\n",
    "# remove RBCs (high hemoglobin gene expression, group 14) and platelets (NRGN, TUBB1, group 9)\n",
    "#adata = adata[~adata.obs[groupby].isin([\"14\", \"9\"])]\n",
    "\n",
    "# rough annotations\n",
    "adata.obs[\"rough_annot\"] = [\"Monocytes\" if c in [\"6\",\"7\",\"10\",\"22\",\"25\",\"28\",\"29\",\"30\",\"33\",\"39\", \"40\"] #CD14+/CD16+ \n",
    "                            else \"CD4Tcells\" if c in [\"1\",\"2\",\"4\",\"8\",\"11\",\"12\",\"13\",\"15\",\"16\",\"19\",\"24\",\"26\",\"27\",\"32\",\"34\",\"36\",\"37\"]\n",
    "                            else \"CD8Tcells\" if c in [\"9\",\"14\"]\n",
    "                            else \"NKcells\" if c in [\"0\", \"23\"]\n",
    "                            else \"Bcells\" if c in [\"3\", \"5\", \"20\",\"35\",\"38\"]  \n",
    "                            else \"Platelets_RBCs\" if c in [\"17\",\"31\",\"41\"]\n",
    "                            else \"Doublets\" if c in [\"18\",\"21\"]\n",
    "                            else \"Dead_cells\" if c in [\"42\", \"43\"]\n",
    "                            else \"NA\" for c in adata.obs[groupby]]\n",
    "\n",
    "sc.pl.umap(adata, color=[\"rough_annot\"])\n",
    "\n",
    "# plot proportion of cells in each sample\n",
    "tmp = pd.crosstab(adata.obs['Sample'],adata.obs['rough_annot'], normalize='index')\n",
    "tmp.plot.bar(stacked=True, figsize=(15,6)).legend(loc='center', bbox_to_anchor=(1, 0., 0.5, 0.5))\n",
    "\n",
    "# Remove dead cells, hb and platelets\n",
    "adata =adata[~(adata.obs[\"rough_annot\"].str.contains(\"Platelets_RBCs|Doublets|Dead_cells\")),:]\n",
    "sc.pl.umap(adata, color=[\"rough_annot\"])\n",
    "\n",
    "adata.uns['log1p']['base'] = None\n",
    "adata.write(outputPath)\n",
    "\n",
    "### Creation of the subset adata\n",
    "for subset in cells_dict:\n",
    "    \n",
    "    seed = 123\n",
    "    np.random.seed(seed)\n",
    "\n",
    "    print(subset)\n",
    "    \n",
    "    # get subset\n",
    "    adata_sub = adata[adata.obs[\"rough_annot\"] == subset]\n",
    "\n",
    "    print(subset)\n",
    "    print(cells_dict[subset][\"leiden\"])\n",
    "\n",
    "    # values for HVG cutoffs values\n",
    "    print(adata_sub.var[\"mean_counts\"].min())\n",
    "    print(adata_sub.var[\"mean_counts\"].mean())\n",
    "    print(adata_sub.var[\"mean_counts\"].max())\n",
    "\n",
    "    ### batch correct using HarmonyPy\n",
    "    # show UMAP before batch correction\n",
    "    # #TODO - check if need to scale/regress out again\n",
    "\n",
    "    #calculateUMAP(adata, flavor='seurat', min_mean=0.0125, max_mean=3, min_disp=0.5, hvg_batch='batch',\n",
    "    #n_neighbors=10, n_pcs=60, return_full=False, umap=False):\n",
    "\n",
    "    adata_umap = calculateUMAP(adata_sub, return_full=False, n_pcs=40,min_mean=0.01, max_mean=4, min_disp=0.5,  \n",
    "                                   hvg_batch=\"batch\", umap=False, n_neighbors=10) \n",
    "\n",
    "    sc.pl.umap(adata_sub, color=['sampleType', 'batch', 'Lane'], wspace=0.35)\n",
    "\n",
    "    # batch correct\n",
    "    pc_df = batchCorrect(adata_umap, output_path=None, max_iter_h=3, vars_use=[\"batch\"])\n",
    "    adata_sub.obsm['X_pca'] = pc_df\n",
    "    sc.pl.umap(adata_sub, color=['sampleType', 'batch', 'Lane'], wspace=0.35)\n",
    "\n",
    "    # calculate updated UMAP\n",
    "    sc.pp.neighbors(adata_sub, n_pcs=40, n_neighbors=10)\n",
    "    sc.tl.umap(adata_sub, min_dist=0.1, spread=1)\n",
    "    sc.pl.umap(adata_sub, color=['dx', 'batch', 'Lane'], wspace=0.35)\n",
    "\n",
    "    #### leiden clustering\n",
    "    sc.tl.leiden(adata_sub, resolution=cells_dict[subset][\"leiden\"], key_added=leiden_key+\"_r\"+str(cells_dict[subset][\"leiden\"]))\n",
    "\n",
    "    # fine annotations\n",
    "    sc.tl.rank_genes_groups(adata_sub, groupby=leiden_key+\"_r\"+str(cells_dict[subset][\"leiden\"]), use_raw=False, pts=True, \n",
    "                            key_added=leiden_key+\"_rank\", method='wilcoxon')\n",
    "\n",
    "    # Embedding density (Activity score)\n",
    "    #sc.tl.embedding_density(adata_sub, basis='umap', groupby=\"activity_python_binary\")\n",
    "\n",
    "    ### save to final file\n",
    "    adata_sub.write(\"./dataOutput/subset_adata/\" + subset +\"_normalized.h5ad\")\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fine annotation of subsets clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### params\n",
    "plt.rcParams['figure.figsize'] = [5, 5]\n",
    "rough_annot = \"rough_annot\"\n",
    "leiden_key = \"leiden_subset\"\n",
    "global_markers_genes = [\"CD2\", \"CD3D\", \"CD8A\",\"CD4\",\"CD33\", \"ITGAM\",\"CD14\", \"FCGR3A\", \"NKG7\", \"NCAM1\", \"CD79A\",  \"CD19\"]\n",
    "hue_groups = [\"control\", \"No\", \"Yes\"] #, [\"Positive\", \"Negative\"] #[\"control\", \"No\", \"Yes\"]\n",
    "hue = \"activity_python_binary_crp\" #\"MY_MTX\" #\"dx\" #\"activity_python_binary\"\n",
    "\n",
    "\n",
    "for subset in cells_dict:\n",
    "    \n",
    "    print(subset)\n",
    "    seed = 123\n",
    "    np.random.seed(seed)\n",
    "    \n",
    "    adata_sub = sc.read_h5ad(\"./dataOutput/subset_adata/\" + subset +\"_normalized.h5ad\")\n",
    "    adata_sub.uns['log1p']['base'] = None\n",
    "    \n",
    "    marker_genes_subset = cells_dict[subset][\"markergenes\"]\n",
    "    \n",
    "    #subset leiden clustering\n",
    "    sc.pl.umap(adata_sub, color=leiden_key+\"_r\"+str(cells_dict[subset][\"leiden\"]))\n",
    "    sc.tl.rank_genes_groups(adata_sub, groupby=leiden_key+\"_r\"+str(cells_dict[subset][\"leiden\"]), use_raw=False, pts=True, key_added=leiden_key+\"_rank\", method='wilcoxon')\n",
    "    sc.pl.rank_genes_groups(adata_sub, n_genes=25, sharey=False, key=leiden_key+\"_rank\", use_raw = False, ncols=6)\n",
    "    sc.pl.rank_genes_groups_dotplot(adata_sub, n_genes=10, key=leiden_key+\"_rank\", groupby=leiden_key+\"_r\"+str(cells_dict[subset][\"leiden\"]), use_raw = False)\n",
    "    \n",
    "    #subset leiden cluster\n",
    "    cluster_small_multiples(adata_sub, clust_key=leiden_key+\"_r\"+str(cells_dict[subset][\"leiden\"]))\n",
    "    \n",
    "    #marker genes\n",
    "    sc.pl.umap(adata_sub, color = global_markers_genes+marker_genes_subset, size = 50)\n",
    "\n",
    "    #Violin plot and  subset cluster quality control\n",
    "    plt.rcParams['figure.figsize'] = [40, 5]\n",
    "    sc.pl.violin(adata_sub, ['n_genes_by_counts'],jitter=0.1, groupby = leiden_key+\"_r\"+str(cells_dict[subset][\"leiden\"]), rotation= 90,cut=0) \n",
    "    sc.pl.violin(adata_sub, ['pct_counts_mt'],jitter=0.1, groupby = leiden_key+\"_r\"+str(cells_dict[subset][\"leiden\"]), rotation= 90,cut=0) \n",
    "    sc.pl.violin(adata_sub, ['pct_counts_ribo'],jitter=0.1, groupby = leiden_key+\"_r\"+str(cells_dict[subset][\"leiden\"]), rotation= 90,cut=0) \n",
    "    sc.pl.violin(adata_sub, ['pct_counts_hb'],jitter=0.1, groupby = leiden_key+\"_r\"+str(cells_dict[subset][\"leiden\"]), rotation= 90,cut=0) \n",
    "    plt.rcParams['figure.figsize'] = [5, 5]\n",
    "    \n",
    "    ##### Subset annotation\n",
    "    adata_sub.obs[\"fine_annot\"] = adata_sub.obs[leiden_key+\"_r\"+str(cells_dict[subset][\"leiden\"])].map(cells_dict[subset][\"fine_annot\"])\n",
    "    sc.pl.umap(adata_sub, color=\"fine_annot\")\n",
    "    \n",
    "    # load color palette\n",
    "    '''colors = {\"Bcells\":(\"Blues\", 3), \"CD4Tcells\":(\"Reds\", 5), \"CD8Tcells\":(\"pink_r\", 4),\n",
    "         \"Monocytes\":(\"Purples\", 5), \"NKcells\":(\"Greens\", 3)}\n",
    "    subset_colors = []\n",
    "\n",
    "    for c in colors.keys():\n",
    "    '''\n",
    "    \n",
    "    #Removing platelettes and doublets\n",
    "    adata_sub =adata_sub[~(adata_sub.obs[\"fine_annot\"].str.contains(\"Dead_cells|Platelets_RBCs|Doublets\")),:]\n",
    "    \n",
    "    adata_sub = adata_sub[:,~adata_sub.var_names.str.startswith('MT-')]\n",
    "    #Should we remove RB genes? in the DEG and in rank genes\n",
    "    adata_sub = adata_sub[:,~adata_sub.var_names.str.startswith('RPL')]\n",
    "    adata_sub = adata_sub[:,~adata_sub.var_names.str.startswith('RPS')]\n",
    "    adata = adata[:,~adata.var_names.str.startswith('MRPL')]\n",
    "    adata = adata[:,~adata.var_names.str.startswith('MRPS')]\n",
    "    \n",
    "    sc.tl.rank_genes_groups(adata_sub, groupby=\"fine_annot\", use_raw=False, pts=True, \n",
    "                        key_added=leiden_key+\"_rank\", method='wilcoxon')\n",
    "    \n",
    "\n",
    "    sc.pl.umap(adata_sub, color=\"fine_annot\", palette = cells_dict[subset][\"color_subset\"])\n",
    "    sc.pl.rank_genes_groups(adata_sub, key=leiden_key+\"_rank\", n_genes=15, sharey=False, use_raw = False, ncols=4 )\n",
    "    sc.pl.rank_genes_groups_dotplot(adata_sub, n_genes=3, key=leiden_key+\"_rank\", groupby=\"fine_annot\", use_raw = False, cmap = cells_dict[subset][\"color_cmap\"])  \n",
    "\n",
    "\n",
    "    ### Subset analysis\n",
    "       \n",
    "    adata_sub.write(\"./dataOutput/subset_adata/\" + subset +\"_clustered.h5ad\")    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  Final adata with subsets clustering "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Merging fine annot\n",
    "plt.rcParams['figure.figsize'] = [5, 5]\n",
    "inputPath = './dataOutput/adata/roughannot.h5ad'\n",
    "outputPath = './dataOutput/adata/final.h5ad'\n",
    "\n",
    "adata = sc.read_h5ad(inputPath)\n",
    "\n",
    "adata = adata[:,~adata.var_names.str.startswith('MT-')]\n",
    "adata = adata[:,~adata.var_names.str.startswith('RPL')]\n",
    "adata = adata[:,~adata.var_names.str.startswith('RPS')]\n",
    "adata = adata[:,~adata.var_names.str.startswith('MRPL')]\n",
    "adata = adata[:,~adata.var_names.str.startswith('MRPS')]\n",
    "    \n",
    "\n",
    "fine_annot_dict = {}\n",
    "\n",
    "for subset in cells_dict:\n",
    "    print(subset)\n",
    "    adata_sub = sc.read_h5ad(\"./dataOutput/subset_adata/\" + subset +\"_clustered.h5ad\")\n",
    "    fine_annot_sub = dict(zip(adata_sub.obs.index, adata_sub.obs[\"fine_annot\"]))\n",
    "    fine_annot_dict.update(fine_annot_sub)\n",
    "\n",
    "adata.obs[\"fine_annot\"] = adata.obs.index.map(fine_annot_dict)\n",
    "\n",
    "# reorder values\n",
    "#adata.obs[fine_annot] = adata.obs[final_annot].astype(\"category\")\n",
    "#adata.obs[fine_annot].cat.reorder_categories(cell_subsets, inplace=True)\n",
    "#sc.pl.umap(adata,color = fine_annot, palette = sns.color_palette(subset_colors), title=\"\", size=3)\n",
    "\n",
    "#adata.obs[\"fine_color\"] = adata.obs.index.map(fine_color_dict)\n",
    "\n",
    "adata = adata[adata.obs[\"fine_annot\"].notnull()]\n",
    "\n",
    "adata.obs[\"fine_annot\"] = adata.obs[\"fine_annot\"].astype(\"category\")\n",
    "\n",
    "adata.obs['fine_annot'].cat.reorder_categories(\n",
    "['CD4 T central memory','CD4 T effector memory','CD4 T IFIT','CD4 T Naive','yd T cells','CD8 T early Tem',\n",
    "'CD8 T Naive','CD8 TEMRA','Classical Monocytes','IFITM3 Monocytes',\n",
    " 'IL1b-Monocytes','Myeloid DCs','Non-classical Monocytes','NKCD56bright','NKCD56low',\n",
    " 'Memory Bcells', 'Naive Bcells','Plasmablasts'], inplace = True)\n",
    "\n",
    "color_subset = ['#ea4e40','#f37761', '#d33028', '#b3302a', '#72322f',\n",
    "                '#FFD6A5','#d08c60','#99582a',\n",
    "                '#afaed4','#8d89c0','#705eaa', '#7209b7','#3c096c',\n",
    "                '#8ed08b', '#2c944c',\n",
    "                '#2a6f97','#4cc9f0','#023e8a','#0267c1',  '#A0C4FF']\n",
    "                \n",
    "sc.pl.umap(adata, color = \"fine_annot\", palette = color_subset)\n",
    "\n",
    "adata.write(outputPath)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
